File: javaAlgorithms/Detection.java
public class Detection {
    private String code;
    private String codeWithErrors;
    private String cleanedData;
    private String errors;
    private String parities;
    private String corrected;

    public Detection(String code, String errors) {
        this.code = code;
        this.codeWithErrors = errors;
    }

    public void decode(Hamming hamming) {
        this.cleanedData = this.cleanBits();
        Hamming h = new Hamming(11, 7, this.cleanedData);
        h.hamming();
        this.errors = this.comparison(hamming.parities, h.parities);
        this.corrected = this.fixErrors(hamming.parities);
    }

    public String cleanBits() {
        String cleaned = "";
        this.parities = "";

        for (int i = 0; i < this.codeWithErrors.length(); i++) {
            if ((i + 1) != Math.pow(2, (int) Math.floor(Math.log(i + 1) / Math.log(2)))){
                cleaned += this.codeWithErrors.charAt(i);
            } else {
                this.parities += this.codeWithErrors.charAt(i);
            }
        }

        return cleaned;
    }

    public String comparison(String lastOne, String withErrors) {
        String errors = "";

        for (int i = 0; i < lastOne.length(); i++) {
            if (lastOne.charAt(i) == withErrors.charAt(i)) {
                errors += "0";
            } else {
                errors += "1";
            }
        }

        return new StringBuilder(errors).reverse().toString();
    }

    public int toDecimal() {
        int d = 0;
        String rev = new StringBuilder(this.errors).reverse().toString();

        for (int i = 0; i < rev.length(); i++) {
            if (rev.charAt(i) == '1') {
                d += (int) Math.pow(2, i);
            }
        }

        return d;
    }

    public String fixErrors(String parities) {
        int count = 0;
        String compared = this.comparison(parities, this.parities);
        StringBuilder result = new StringBuilder();
        int index = -1;

        for (int i = 0; i < this.errors.length(); i++) {
            if (this.errors.charAt(i) == '1') {
                count++;
            }
        }

        for (int i = 0; i < compared.length(); i++) {
            if (this.errors.charAt(i) == '1') {
                count++;
                index = i;
            }
        }

        if (count == 0) {
            result.append(this.codeWithErrors);
        } else if (count == 1) {
            int errorPos = (int) Math.pow(2, index);
            char val = this.codeWithErrors.charAt(errorPos);
            char y = val == '0' ? '1' : '0';
            result.append(this.codeWithErrors, 0, errorPos - 1).append(y).append(this.codeWithErrors.substring(errorPos));
        } else {
            int p = this.toDecimal();
            char val = this.codeWithErrors.charAt(p - 1);
            char y = val == '0' ? '1' : '0';
            result.append(this.codeWithErrors, 0, p - 1).append(y).append(this.codeWithErrors.substring(p));
        }

        return result.toString();
    }

    public void printResults() {
        System.out.println("Deteccion:\n" +
                "Original: " + this.codeWithErrors + "\n" +
                "Corregida: " + this.corrected + "\n");
    }
}

File: pythonAlgorithms/detection.py
from math import floor, log, pow
from hamming import Hamming

class Detection:
    
    def __init__(self, code, errors) -> None:
        self.code = code
        self.code_with_errors = errors
    


    def decode(self, hamming:Hamming):
        self.cleaned_data = self.clean_bits()
        h = Hamming(11,7,self.cleaned_data)
        h.hamming()
        self.errors = self.comparison(hamming.parities,h.parities)
        self.corrected = self.fix_errors(hamming.parities) 
    


    def clean_bits(self):
        cleaned = ""
        self.parities = ""

        for i in range(len(self.code_with_errors)):
            if (i + 1) != pow(2, int( floor( log(i + 1) / log(2) ) ) ):
                cleaned += self.code_with_errors[i]
            else:
                self.parities += self.code_with_errors[i]

        return cleaned
    

    
    def comparison(self, last_one, with_errors):
        errors = ""

        for i in range(len(last_one)):
            if last_one[i] == with_errors[i]:
                errors += "0"
            else:
                errors += "1"
        
        return "".join(reversed(errors))



    def to_decimal(self):
        d = 0
        rev = self.errors[::-1]

        for i in range(len(rev)):
            if rev[i] == '1':
                d += 2**i

        return d
    


    def fix_errors(self, parities):
        count = 0
        compared = self.comparison(parities, self.parities)[::-1]
        result = ""
        index = -1

        for i in range(len(self.errors)):
            if self.errors[i] == "1":
                count += 1
        
        for i in range(len(compared)):
            if self.errors[i] == "1":
                count += 1
                index = i
        
        if count == 0:    
            result = self.code_with_errors

        elif count == 1:
            error_pos = 2**index
            val = self.code_with_errors[error_pos]
            result = self.code_with_errors[:self.code_with_errors-1] + ('1' if val == '0' else '0') + self.code_with_errors[error_pos:]

        else:
            p = self.to_decimal()
            val = self.code_with_errors[p - 1]
            result = self.code_with_errors[:p-1] + ("1" if val == "0" else "0") + self.code_with_errors[p:]
        
        return result
    


    def print_results(self):
        print(f"""Deteccion:
        Original: {self.code_with_errors}
        Corregida: {self.corrected}
        """)
    